sql/optbuilder/builder.go

```diff
@@ -0,0 +1,1309 @@
+package optbuilder
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"strings"
+
+	ast "github.com/dolthub/vitess/go/vt/sqlparser"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/encodings"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/expression/function"
+	"github.com/dolthub/go-mysql-server/sql/expression/function/aggregation"
+	"github.com/dolthub/go-mysql-server/sql/mysql_db"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+type scope struct {
+	b      *PlanBuilder
+	parent *scope
+	ast    ast.SQLNode
+	node   sql.Node
+
+	cols      []scopeColumn
+	extraCols []scopeColumn
+	ctes      map[string]*scope
+	groupBy   *groupBy
+	exprs     map[string]columnId
+	colId     columnId
+}
+
+func (s *scope) getExpr(name string) (columnId, bool) {
+	n := strings.ToLower(name)
+	id, ok := s.exprs[n]
+	if !ok && s.groupBy != nil {
+		id, ok = s.groupBy.outScope.getExpr(n)
+	}
+	return id, ok
+}
+
+func (s *scope) initGroupBy() {
+	s.groupBy = &groupBy{outScope: s.replace()}
+}
+
+func (s *scope) outerScopeLen() int {
+	var cnt int
+	sco := s.parent
+	for sco != nil {
+		cnt += len(sco.cols)
+		sco = sco.parent
+	}
+	return cnt
+}
+
+func (s *scope) setTableAlias(t string) {
+	for i := range s.cols {
+		beforeColStr := s.cols[i].String()
+		s.cols[i].table = t
+		id, ok := s.getExpr(beforeColStr)
+		if !ok {
+			err := sql.ErrColumnNotFound.New(beforeColStr)
+			s.b.handleErr(err)
+		}
+		delete(s.exprs, beforeColStr)
+		s.exprs[s.cols[i].String()] = id
+	}
+}
+
+func (s *scope) setColAlias(cols []string) {
+	if len(cols) != len(s.cols) {
+		s.b.handleErr(fmt.Errorf("invalid column number for column alias"))
+	}
+	for i := range s.cols {
+		s.cols[i].col = cols[i]
+	}
+}
+
+func (s *scope) push() *scope {
+	return &scope{
+		parent: s,
+	}
+}
+
+func (s *scope) replace() *scope {
```

Comment:
This is a weird operation, only copying over a couple fields of the original. What's it for?

Reply:
Used for sibling scopes. For example aggregation outputs, projection outputs, CTE outputs. Each has the same parent as the fromScope they are copied from. All of these are copied and populated with a set of columns that are used for resolving in certain situations. When I do more with subqueries and CTEs I might notice the parent reference not being necessary in some cases.